clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java

```diff
@@ -145,6 +148,39 @@ public void testResetPositionsProcess() {
         verify(applicationEventProcessor).process(any(ResetPositionsEvent.class));
     }
 
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testAssignmentChangeEvent(boolean withGroupId) {
+        final long currentTimeMs = 12345;
+        AssignmentChangeEvent event = new AssignmentChangeEvent(currentTimeMs, 12345, Collections.emptyList());
+
+        setupProcessor(withGroupId);
+        doReturn(true).when(subscriptionState).assignFromUser(any());
+        processor.process(event);
+        if (withGroupId) {
+            verify(commitRequestManager).updateAutoCommitTimer(currentTimeMs);
+            verify(commitRequestManager).maybeAutoCommitAsync();
+        } else {
+            verify(commitRequestManager, never()).updateAutoCommitTimer(currentTimeMs);
+            verify(commitRequestManager, never()).maybeAutoCommitAsync();
+        }
+        verify(metadata).requestUpdateForNewTopics();
+        assertDoesNotThrow(() -> event.future().get());
+
+    }
+
+    @Test
+    public void testAssignmentChangeEventWithException() {
+        AssignmentChangeEvent event = new AssignmentChangeEvent(12345, 12345, Collections.emptyList());
+
+        setupProcessor(false);
+        doThrow(new IllegalStateException()).when(subscriptionState).assignFromUser(any());
+        processor.process(event);
+
+        ExecutionException e = assertThrows(ExecutionException.class, () -> event.future().get());
+        assertInstanceOf(IllegalStateException.class, e.getCause());
```

Comment:
Where is the `IllegalStateException` thrown in a production setting? I see we special-cased `IllegalStateException` in `ConsumerUtils`, but I'm wondering if an `IllegalStateException` is something a user can hit 🤔

Reply:
This is in `SubscriptionState`.

https://github.com/apache/kafka/blob/42715654095d019effc478ab343c0132e02aa70f/clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java#L230-L231

In `KafkaConsumer#assign` java doc, we also mention this exception.

https://github.com/apache/kafka/blob/42715654095d019effc478ab343c0132e02aa70f/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L783-L784